### PR TITLE
Accomodate Arbitrary Platform Reserved Addresses

### PR DESCRIPTION
**Pull request by iamalwaysuncomfortable**
### Motivation/In This PR

The motivation for [MCIP #36: Reserved Subaddresses](https://github.com/mobilecoinfoundation/mcips/blob/main/text/0036-reserved-subaddresses.md) states:
_In developing new features, we have discovered that sometimes, we want to reserve particular subaddress indices for specific purposes. This is particularly helpful for mobile wallets which are designed to be used with "linked devices" that share your private keys. By using the subaddress index to represent information about a TxOut, they can ensure that other linked devices will all see this information, without needing to create another synchronization mechnaism._

These addresses are index along the following scheme: 
| Subaddress index | Purpose         |
| ---------------- | --------------- |
| `u64::MAX`       | Invalid / None      |
| `u64::MAX - 1`   | Change subaddress [MCIP #4](https://github.com/mobilecoinfoundation/mcips/pull/4)       |
| `u64::MAX - 2`   | Gift code subaddress [MCIP #32](https://github.com/mobilecoinfoundation/mcips/pull/32) |

The supporting `ChangeAddress` object provides data fields for storing the actual derived addresses corresponding to the change subaddress and primary addresses. Given the list of official reserved subaddresses has grown, this PR:

* Expands the `ChangeAddress` object to hold all of the reserved sub-addresses and is renamed to `ReservedAddress`.  
* Adds the gift code reserved subaddresses specified in MCIPs 32 & 36

This PR addresses issue: #1890

### CODE REVIEW

File: android-bindings/src/bindings.rs [hunk 1]
```diff
@@ -1747,8 +1747,8 @@ pub unsafe extern "C" fn Java_com_mobilecoin_lib_TransactionBuilder_add_1change_
                 env.get_rust_field(source_account_key, RUST_OBJ_FIELD)?;
 
             let value = jni_big_int_to_u64(env, value)?;
-            let change_destination: ChangeDestination =
-                ChangeDestination::from(&*source_account_key);
+            let change_destination: ReservedDestination =
+                ReservedDestination::from(&*source_account_key);
```
Comment (remoun) [inline android-bindings/src/bindings.rs:1751]: Nit: DRY
```suggestion
            let change_destination = ReservedDestination::from(&*source_account_key);
```

File: transaction/std/src/reserved_destination.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        change_subaddress_index: Option<u64>,
+        gift_code_subaddress_index: Option<u64>,
+    ) -> Self {
+        let change_subaddress = if let Some(change_subaddress) = change_subaddress_index {
+            acct.subaddress(change_subaddress)
+        } else {
+            acct.change_subaddress()
+        };
+
+        let gift_code_subaddress = if let Some(gift_code_subaddress) = gift_code_subaddress_index {
```
Comment (remoun) [inline transaction/std/src/reserved_destination.rs:61]: Optional/Equivalent: Use `Into<Option<u64>>` to allow callers to omit `Some()`

```suggestion
        // allows callers to pass `n`, `Some(n)` or `None`
        change_subaddress_index: impl Into<Option<u64>>,
        gift_code_subaddress_index: impl Into<Option<u64>>,
    ) -> Self {
        let change_subaddress = if let Some(change_subaddress) = change_subaddress_index.into() {
            acct.subaddress(change_subaddress)
        } else {
            acct.change_subaddress()
        };

        let gift_code_subaddress = if let Some(gift_code_subaddress) = gift_code_subaddress_index.into() {
```

**Review by remoun — APPROVED** [review]
LGTM

File: transaction/std/src/reserved_destination.rs (hunk 2 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/reserved_destination.rs:61]: Good suggestion. Let me think on this one, given the sensitivity/implications of the systems this touches, I think I may want to keep the Some() to make it explicit to the caller that they're making an active change to system defaults.

File: transaction/std/src/reserved_destination.rs [hunk 3]
```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Object representing a destination for change from a transaction
```
Comment (nick-mobilecoin) [inline transaction/std/src/reserved_destination.rs:3]: don't have a good suggestion, but looks like there is a stray use of "change" from the rename

File: transaction/std/src/reserved_destination.rs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    /// account. It is the account to which all change outputs are actually
+    /// sent. The account owner is able to confirm that an output is change
+    /// by checking that it matches to the change subaddress.
+    /// This should always be the change_subaddress for some AccountKey.
+    pub change_subaddress: PublicAddress,
+
+    /// A secret reserved subaddress to which gift code ouputs are sent
+    /// when they are created. Similarly to the change_subaddress, gift code
+    /// creators can check to see gift codes they've created by checking outputs
+    /// at this address
```
Comment (nick-mobilecoin) [inline transaction/std/src/reserved_destination.rs:33]: nit
```suggestion
    /// at this subaddress
```

**Review by nick-mobilecoin — APPROVED** [review]

File: transaction/std/src/reserved_destination.rs (hunk 3 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/reserved_destination.rs:3]:
```suggestion
//! Object containing the public destination addresses of MobileCoin standard reserved subaddresses
```

File: transaction/std/src/reserved_destination.rs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            primary_address: src.default_subaddress(),
+            change_subaddress: src.change_subaddress(),
+            gift_code_subaddress: src.gift_code_subaddress(),
+        }
+    }
+}
+
+impl ReservedDestination {
+    /// Set alternate subaddresseses for reserved addresses. This is useful in some things
+    /// like mobilecoind
```
Comment (iamalwaysuncomfortable) [inline transaction/std/src/reserved_destination.rs:49]:
```suggestion
    /// Set alternate subaddresseses for reserved addresses. This is useful in
    /// some things like mobilecoind
```

File: android-bindings/src/bindings.rs (hunk 1 above)
Comment (iamalwaysuncomfortable) [reply android-bindings/src/bindings.rs]: Changed!

File: transaction/std/src/reserved_destination.rs (hunk 3 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/reserved_destination.rs]: Indeed, addressed!

File: transaction/std/src/reserved_destination.rs (hunk 4 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/reserved_destination.rs]: Added!

**Pull request merged**